klib/gcp.c

```diff
@@ -698,6 +702,51 @@ static void gcp_metrics_add_memory(buffer body, boolean first, const char *type,
                   first ? "" : ",", gcp.zone, gcp.instance_id, type, state, value, interval);
 }
 
+static void gcp_metrics_add_disk(buffer body, boolean first, const char *type, const char *state,
+                                   u64 value, const char *interval, buffer device)
+{
+    bprintf(body, "%s{\"resource\":{\"type\":\"gce_instance\""
+                  ",\"labels\":{\"zone\":\"%b\",\"instance_id\":\"%b\"}}"
+                  ",\"metric\":{\"type\":\"agent.googleapis.com/disk/%s\""
+                  ",\"labels\":{\"device\":\"%b\",\"state\":\"%s\"}}"
+                  ",\"points\":[{\"value\":{\"doubleValue\":%ld},\"interval\":%s}]}",
+                  first ? "" : ",", gcp.zone, gcp.instance_id, type, device, state, value, interval);
+}
+
+closure_function(3, 4, void, gcp_metrics_disk_vh,
+                 buffer, b, const char *, interval, boolean, include_readonly,
+                 u8 *, uuid, const char *, label, filesystem, fs, inode, mount_point)
+{
+    if (filesystem_is_readonly(fs) && !bound(include_readonly))
+        return;
+    int block_size = fs_blocksize(fs);
+    u64 total_blocks = fs_totalblocks(fs);
+    u64 free_blocks = fs_freeblocks(fs);
+    u64 percent_used_free = free_blocks * 100 / total_blocks;
+    buffer b = bound(b);
+    const char *interval = bound(interval);
+    buffer device = NULL;
+    if (label[0]) {
+        device = alloca_wrap_cstring(label);
+    } else if (uuid) {
+        buffer device = little_stack_buffer(2 * UUID_LEN + 4);
```

Comment:
This is creating a new `device` local variable whose scope is limited to the `else if` block, instead of assigning a value to the outer `device` variable. It should be `device = little_stack_buffer(2 * UUID_LEN + 4);`
Also, no need for the `if (uuid)` part, and no need to initialize device to NULL (when the label is empty, the uuid is never null, and we cannot pass a null buffer to bprintf() anyway).